### PR TITLE
[JN-971] Search exp query builder

### PR DESCRIPTION
**Pull request by connorlbark**
#### DESCRIPTION

Adds a prototype-user-only search query builder and new endpoint which returns all of the possible facets with their types. Using react-querybuilder, we're able to create new expressions, but we're not able to parse existing ones. That will be the subject of a future PR, likely using Antlr.

#### TO TEST:  *(simple manual steps for confirming core behavior -- used for pre-release checks)*

- Populate ourhealth
- go to the modal to create a new survey
- Test out the query builder

### CODE REVIEW

File: api-admin/src/main/java/bio/terra/pearl/api/admin/controller/search/EnrolleeSearchExpressionController.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+  public EnrolleeSearchExpressionController(
+      EnrolleeSearchExpressionService enrolleeSearchExpressionService,
+      AuthUtilService authUtilService,
+      HttpServletRequest request) {
+    this.enrolleeSearchExpressionService = enrolleeSearchExpressionService;
+    this.authUtilService = authUtilService;
+    this.request = request;
+  }
+
+  @Override
```
Comment (devonbush) [inline api-admin/src/main/java/bio/terra/pearl/api/admin/controller/search/EnrolleeSearchExpressionController.java]: how about we just make this "getExpressionSearchFacets" and put this in EnrolleeSearchController.  That will make it easier to see/compare it side-by-side with the existing getSearchFacets() path, so we don't forget about either path

File: api-admin/src/main/java/bio/terra/pearl/api/admin/controller/search/EnrolleeSearchExpressionController.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  public ResponseEntity<Object> getSearchFacets(
+      String portalShortcode, String studyShortcode, String envName) {
+    AdminUser user = authUtilService.requireAdminUser(request);
+    authUtilService.authUserToStudy(user, portalShortcode, studyShortcode);
```
Comment (devonbush) [inline api-admin/src/main/java/bio/terra/pearl/api/admin/controller/search/EnrolleeSearchExpressionController.java]: how about we make a new EnrolleeSearchExtService to handle the auth.  The auth rules for search are going to get complicated (e.g. some users will be able to see PHI, some wont).  Then you can also relocate the existing EnrolleeExtService.search method there

File: core/src/main/java/bio/terra/pearl/core/service/search/EnrolleeSearchExpressionService.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import bio.terra.pearl.core.service.search.terms.SearchValue;
+import bio.terra.pearl.core.service.survey.SurveyService;
+import org.springframework.stereotype.Service;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@Service
+public class EnrolleeSearchExpressionService {
```
Comment (devonbush) [inline core/src/main/java/bio/terra/pearl/core/service/search/EnrolleeSearchExpressionService.java]: let's move these methods inside EnrolleeSearchService

File: ui-admin/src/api/api.tsx [hunk 4]
```diff
@@ -789,6 +791,13 @@ export default {
     return await this.processJsonResponse(response)
   },
 
+  async getSearchFacetsV2(portalShortcode: string, studyShortcode: string, envName: string):
```
Comment (devonbush) [inline ui-admin/src/api/api.tsx]:
```suggestion
  async getExpressionSearchFacets(portalShortcode: string, studyShortcode: string, envName: string):
```

**Review by devonbush — CHANGES_REQUESTED** [review]
Overall looks good.  My main feedback is to put the new search methods side-by-side with the old, rather than building a parallel class hierarchy.  That will make it easier to compare/contrast functionality, and to clean out old code once it's no longer needed

File: api-admin/src/main/java/bio/terra/pearl/api/admin/controller/search/EnrolleeSearchExpressionController.java (hunk 1 above)
Comment (connorlbark) [reply api-admin/src/main/java/bio/terra/pearl/api/admin/controller/search/EnrolleeSearchExpressionController.java]: I was doing it this way to make it easier to delete the code in one big fell swoop later / make it more clear what code is deprecated versus new, but I could go either way. Happy to move that over

File: ui-admin/src/util/formatQueryBuilderAsSearchExp.tsx [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+/**
+ * Heavily lifted from react-querybuilder's SpEL rule processor, but with some modifications
+ * to make it work with our search format. Not all of these features are supported in our
+ * search format, but they are here for completeness.
+ */
+export const ruleProcessorEnrolleeSearchExpression: RuleProcessor = (
+  { field, operator, value, valueSource },
+  { escapeQuotes, parseNumbers } = {}
+) => {
+  console.log('called')
```
Comment (MatthewBemis) [inline ui-admin/src/util/formatQueryBuilderAsSearchExp.tsx]:
```suggestion
```

File: core/src/main/java/bio/terra/pearl/core/service/participant/search/EnrolleeSearchService.java [hunk 6]
```diff
@@ -62,4 +73,24 @@ protected EnrolleeSearchFacet getTaskFacet(StudyEnvironment studyEnvironment) {
...
         return tasksFacet;
     }
+
+    public Map<String, SearchValue.SearchValueType> getExpressionSearchFacetsForStudyEnv(UUID studyEnvId) {
+
+        Map<String, SearchValue.SearchValueType> fields = new HashMap<>();
+        // profile fields
+        ProfileTerm.FIELDS.forEach((term, type) -> fields.put("profile." + term, type));
+        // age
+        fields.put("age", SearchValue.SearchValueType.INTEGER);
```
Comment (MatthewBemis) [inline core/src/main/java/bio/terra/pearl/core/service/participant/search/EnrolleeSearchService.java:83]: While testing locally, I'm getting an NPE when trying to save a survey with an expression that uses the `age` facet:

```
java.lang.NullPointerException: temporal
	at java.base/java.util.Objects.requireNonNull(Objects.java:235)
	at java.base/java.time.LocalDate.from(LocalDate.java:394)
	at java.base/java.time.LocalDate.until(LocalDate.java:1702)
	at java.base/java.time.Period.between(Period.java:389)
	at bio.terra.pearl.core.service.search.terms.AgeTerm.extract(AgeTerm.java:28)
	at bio.terra.pearl.core.service.search.expressions.EnrolleeTermComparisonFacet.evaluate(EnrolleeTermComparisonFacet.java:40)
	...
```

File: ui-admin/src/search/SearchQueryBuilder.tsx [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+      controlClassnames={{
+        fields: 'form-select',
+        value: 'form-control',
+        operators: 'form-select',
+        removeRule: 'btn btn-outline-dark',
+        removeGroup: 'btn btn-outline-dark',
+        addRule: 'btn btn-outline-dark',
+        addGroup: 'btn btn-outline-dark d-none',
+        combinators: 'form-select w-25'
+      }}
```
Comment (MatthewBemis) [inline ui-admin/src/search/SearchQueryBuilder.tsx:61]: Nice, thanks for restyling everything! The original PoC was certainly ugly

File: ui-admin/src/search/SearchQueryBuilder.tsx [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+const CustomFieldSelector = (props: FieldSelectorProps) => {
+  const options = props.options.map(option => {
+    return { label: option.label, value: option.label }
+  })
+
+  return <div className="d w-100" key='test' id={'test'}>
+    <Select
+      key={'test2'} id={'test2'}
```
Comment (MatthewBemis) [inline ui-admin/src/search/SearchQueryBuilder.tsx]: Are these "test" keys/ids leftovers from iterating during development?

**Review by MatthewBemis — COMMENTED** [review]
Just a few nits/testing notes

File: ui-admin/src/search/SearchQueryBuilder.tsx (hunk 8 above)
Comment (connorlbark) [reply ui-admin/src/search/SearchQueryBuilder.tsx]: Oh man I accidentally left a lot in that was my bad, thanks for catching all that.

**Review by MatthewBemis — APPROVED** [review]

**Review by devonbush — APPROVED** [review]

**Pull request merged**